hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -0,0 +1,555 @@
+        case Yield(innElt, innS1) => k(Yield(innElt, (outS, innS1)))
+      }
+    }
+  }
+
+  def leftJoinRightDistinct[P, A, B: ParameterPack](
+    left: Parameterized[P, A],
+    right: Parameterized[P, B],
+    rNil: B,
+    comp: (A, B) => Code[Int]
+  ): Parameterized[P, (A, B)] = new Parameterized[P, (A, B)] {
+    implicit val lsP = left.stateP
+    implicit val rsP = right.stateP
+    type S = (left.S, right.S, (B, Code[Boolean]))
+    val stateP: ParameterPack[S] = implicitly
+    def emptyState = (left.emptyState, right.emptyState, (rNil, false))
+    def length(s0: S): Option[Code[Int]] = left.length(s0._1)
+
+    def init(mb: MethodBuilder, jb: JoinPointBuilder, param: P)(
+      k: Init[S] => Code[Ctrl]
+    ): Code[Ctrl] = {
+      val missing = jb.joinPoint()
+      missing.define { _ => k(Missing) }
+      left.init(mb, jb, param) {
+        case Missing => missing(())
+        case Start(lS) => right.init(mb, jb, param) {
+          case Missing => missing(())
+          case Start(rS) => k(Start((lS, rS, (rNil, false))))
+        }
+      }
+    }
+
+    def step(mb: MethodBuilder, jb: JoinPointBuilder, state: S)(
+      k: Step[(A, B), S] => Code[Ctrl]
+    ): Code[Ctrl] = {
+      val (lS0, rS0, (rPrev, somePrev)) = state
+      left.step(mb, jb, lS0) {
+        case EOS => k(EOS)
+        case Skip(lS) => k(Skip((lS, rS0, (rPrev, somePrev))))
+        case Yield(lElt, lS) =>
+          val push = jb.joinPoint[(B, right.S, (B, Code[Boolean]))](mb)
+          val pull = jb.joinPoint[right.S](mb)
+          val compare = jb.joinPoint[(B, right.S)](mb)
+          push.define { case (rElt, rS, rPrevOpt) =>
+            k(Yield((lElt, rElt), (lS, rS, rPrevOpt)))
+          }
+          pull.define(right.step(mb, jb, _) {
+            case EOS => push((rNil, right.emptyState, (rNil, false)))
+            case Skip(rS) => pull(rS)
+            case Yield(rElt, rS) => compare((rElt, rS))
+          })
+          compare.define { case (rElt, rS) =>
+            ParameterPack.localMemoize(mb, comp(lElt, rElt)) { c =>
+              (c > 0).mux(
+                pull(rS),
+                (c < 0).mux(
+                  push((rNil, rS, (rElt, true))),
+                  push((rElt, rS, (rElt, true)))))
+            }
+          }
+          somePrev.mux(
+            compare((rPrev, rS0)),
+            pull(rS0))
+      }
+    }
+  }
+
+  private[ir] def apply(
+    emitter: Emit,
+    streamIR0: IR,
+    env0: Emit.E,
+    rvas: Emit.RVAS,
+    er: EmitRegion,
+    container: Option[AggContainer]
+  ): EmitStream = {
+    val fb = emitter.mb.fb
+
+    def emitIR(ir: IR, env: Emit.E): EmitTriplet =
+      emitter.emit(ir, env, rvas, er, container)
+
+    def emitStream(streamIR: IR, env: Emit.E): Parameterized[Any, EmitTriplet] =
+      emitPStream[Any](streamIR, env, _ => Code._empty)
+
+    def emitPStream[E](streamIR: IR, env: Emit.E, setupEnv: E => Code[Unit]): Parameterized[E, EmitTriplet] =
```

Comment:
You used to have a contravariant `map` over the parameter type of `Parameterized`. Am I right that `env` and `setupEnv` replace that? If so, do you remember why you got rid of the parameter `map`?

Reply:
i realize now that i can bring back contravariant map in a way that works well, since i've realized how to solve some of the problems i had with it before. here's a commit that brings it back and removes `setupEnv`:

https://github.com/iitalics/hail/commit/73047ffa61a6007621f440962c2ed346b664a800

`guardParam` is a contravariant map on the parameter which can fail; if it fails the stream is `Missing` (this is a common pattern)

let me know if you think i should use this

Reply:
Nice! Something like `guardPararm` is exactly what I was hoping for. I do like this better.